### ISSUE TITLE
Extracting exclude or suppression from configuration

### ISSUE THREAD

**Issue by EelcoLabordus** [microsoft/PSRule#1432]
I'm trying to extract the exclude or suppression from the configuration configuration. I found a way to generate a warning when a suppression is configured. But when a rule is excluded i cannot find it back in any logging.

I am building an Azure pipeline that will validate the configuration, but also display what rules/files are excluded/suppressed.

I have tried reading the ps-rule.yaml by using powershell-yaml but then i get a dotnet error because both use different configurations.

Any ideas how i can solve this problem?

**Comment by BernieWhite** [comment]
@EelcoLabordus Thanks for the question.

You are correct we don't generate a warning for excluded rules but do for suppression. We could do that better, providing an option to turn this on is a good idea.

In terms of how could you do that now. You could use create a custom convention in PSRule to check the configuration. This will get you 99% of the way there as there is a few edge cases that would be hard to cover today.

A convention an extensibility point within PSRule. The way it works is you define a convention in `.Rule.ps1` extension included in a PSRule path (`.ps-rule/` by default).

For example:

```powershell
# Synopsis: Generate a warning for any rule exclusions set by options.
Export-PSRuleConvention 'FlagExclusions' -Initialize {
    $exclusions = (New-PSRuleOption).Rule.Exclude;
    foreach ($exclude in $exclusions) {
        Write-Warning "The following rule was excluded: $exclude";
    }
}
```

We can use the `New-PSRuleOption` cmdlet to read from `ps-rule.yaml` but also capture any environment variables that are set. Reading the YAML of `ps-rule.yaml` wouldn't capture any environment variables. Also if you wanted to do this check outside of PSRule then you could also use the `New-PSRuleOption` to get PSRule options as an object, that way you don't need to read the YAML file with `powershell-yaml`.

To use the convention add the convention to the [Convention.Include](https://microsoft.github.io/PSRule/v2/concepts/PSRule/en-US/about_PSRule_Options/#conventioninclude) option.

For example:

```yaml
convention:
  include:
  - FlagExclusions
```

---

Edge cases to note:

- PSRule supports baselines that can specify the `Rule.Exclude` option, these wouldn't be captured if a baseline is used.
- You can also specify `-Option` on a `Assert-PSRule` or `Invoke-PSRule` command line, which could also specify a `Rule.Exclude` option via a hashtable such as `-Option @{ 'Rule.Exclude' = @() }`.

---

I hope that helps as a workaround. But you can leave this issue open and we can add this to the backlog.

**Comment by BernieWhite** [comment]
@EelcoLabordus As I understood the request. Generating a warning for any excluded rules would meet your requirements.

Is this correct?
